Source : https://docs.sympy.org/latest/tutorials/intro-tutorial/index.html#intro-tutorial

Symbolic computation deals with the computation of mathematical objects symbolically. This means that the mathematical objects are represented exactly, not approximately, and mathematical expressions with unevaluated variables are left in symbolic form.

Let’s take an example. Say we wanted to use the built-in Python functions to compute square roots. We might do something like this

In [2]:
import math 
q = math.sqrt(25)
q

5.0

25 is a perfect square, so we got the exact answer, 5. But suppose we computed the square root of a number that isn’t a perfect square

In [3]:
math.sqrt(8)


2.8284271247461903

Here we got an approximate result. 2.82842712475 is not the exact square root of 8 (indeed, the actual square root of 8 cannot be represented by a finite decimal, since it is an irrational number). If all we cared about was the decimal form of the square root of 8, we would be done.

But suppose we want to go further. Recall that  $$\sqrt{8} = \sqrt{2 \cdot 4} = 2\sqrt{2}$$. We would have a hard time deducing this from the above result. This is where symbolic computation comes in. With a symbolic computation system like SymPy, square roots of numbers that are not perfect squares are left unevaluated by default

In [4]:
math.sqrt(3)

1.7320508075688772

Furthermore—and this is where we start to see the real power of symbolic computation—symbolic results can be symbolically simplified.

## A More Interesting Example
The above example starts to show how we can manipulate irrational numbers exactly using SymPy. But it is much more powerful than that. Symbolic computation systems (which by the way, are also often called computer algebra systems, or just CASs) such as SymPy are capable of computing symbolic expressions with variables.

As we will see later, in SymPy, variables are defined using symbols. Unlike many symbolic manipulation systems, variables in SymPy must be defined before they are used (the reason for this will be discussed in the next section).

Let us define a symbolic expression, representing the mathematical expression $$ x + 2\cdot y $$

In [5]:
from sympy import symbols 
x,y = symbols('x y') 
expr = x + 2*y 
expr

x + 2*y

Note that we wrote x + 2*y just as we would if x and y were ordinary Python variables. But in this case, instead of evaluating to something, the expression remains as just x + 2*y. Now let us play around with it:

In [6]:
expr + 1

x + 2*y + 1

In [7]:
expr - x

2*y

Notice something in the above example. When we typed expr - x, we did not get x + 2*y - x, but rather just 2*y. The x and the -x automatically canceled one another. This is similar to how sqrt(8) automatically turned into 2*sqrt(2) above. This isn’t always the case in SymPy, however:

Here, we might have expected $$ x(x+2y) $$ to transform into $$ x^2 + 2xy $$  but instead we see that the expression was left alone. This is a common theme in SymPy. Aside from obvious simplifications like $$ x -x = 0 $$ and $$ \sqrt{8} = 2\sqrt{2} $$  
 most simplifications are not performed automatically. This is because we might prefer the factored form $$ x(x+2y)$$, or we might prefer the expanded form $$ x^2 + 2xy $$. Both forms are useful in different circumstances. In SymPy, there are functions to go from one form to the other

In [14]:
from sympy import expand , factor 
expr = x**2 + 2*x*y


In [15]:
expand_expr = expand(expr) 
expand_expr

x**2 + 2*x*y

In [16]:
factor_expr = factor(expr)
factor_expr

x*(x + 2*y)